Refactor

theron-wang opened the pull request:
Includes the refactored algorithm for fully solvable projects. At the moment, I removed the modularity model code, but I'll add that back in before the end. I also don't have file output code yet, since I'm waiting to finish unsolved symbol generation (with the new alternates option).

Most of the code in `TargetMemberFinderVisitor` is unchanged from before; I just removed some methods that I don't think are necessary in the new implementation. Most of the new logic is located in `Slicer` and `TypeRuleDependencyMap`.

theron-wang committed:
upgrade to latest versions

theron-wang committed:
Refactor without unsolved symbol generation

theron-wang committed:
make SpeciminRunner runnable

kelloggm commented on src/main/java/org/checkerframework/specimin/JavaParserUtil.java:
```diff
@@ -222,14 +210,31 @@ public static String getValueTypeFromAnnotationExpression(Expression value) {
+   * @return the nearest enclosing class-like node
+   */
+  public static @Nullable TypeDeclaration<?> getEnclosingClassLikeOptional(Node node) {
+    Optional<Node> parent = node.getParentNode();
```
Rather than two copies of code that logically does the same thing, could you implement this version in terms of the other (e.g., by catching an exception) or vice-versa? I think it should be relatively easy to do so.

kelloggm commented on src/main/java/org/checkerframework/specimin/Slicer.java:
```diff
@@ -0,0 +1,137 @@
+
+  /**
+   * The slice of unsolved symbol definitions. These values need not be unique; the map is provided
+   * for simple lookups when adding new symbols.
```
I'm concerned that these values don't need to be unique - maps don't really make sense unless the keys aren't going to overlap. Moreover, I'd generally expect that unsolved symbol names are unique (at least, within a single compilation unit): if not, the original Java code wouldn't compile, either - at least, assuming that the keys here are fully-qualified names. Are they?

kelloggm replied:
> However, using sets as the keys would prompt a new SomeClass, with only FQN org.example.SomeClass, to be generated

I don't think this is necessarily a problem - we could replace the key with an updated one when we discover this constraint, no?

> This is an edge edge case, so we may never encounter this. Do you think it's worth accounting for?

I think it's a good idea to try to get this right the first time, since we're rewriting the core of the analysis. So, let's try to make sure we've thought of and accounted for all of the edge cases.

kelloggm replied:
> My understanding is that you're suggesting a map of type Map<Set<String>, UnsolvedSymbol>, right?

I was imagining a custom data structure for the key that's backed by a string set but has its own looser equality method. Regardless, I think we're on the same page with respect to what this needs to handle; the exact way that you implement it is less important. You should do what you find most intuitive.

kelloggm commented on src/main/java/org/checkerframework/specimin/Slicer.java:
```diff
@@ -0,0 +1,137 @@
+      slice.add(node);
+    }
+
+    worklist.addAll(TypeRuleDependencyMap.getRelevantElements(node));
```
The type rule dependency map will eventually need to be an object that is passed in, I think.

theron-wang committed:
unsolved symbol generation part 1

theron-wang committed:
forgot to include changes to `Slicer`

theron-wang commented:
@kelloggm I just pushed some commits for unsolved symbol generation. At this point, I haven't added support for type parameters and some more complex inference logic yet.

I'm aiming to make `UnsolvedSymbolGenerator.addInformation` into a method that can be called multiple times on the same generated symbol, each in different contexts. For example, `inferContextImpl` will generate the symbols, but `addInformation` will see if anything needs to be changed from the original definition. If a type is found in a `try` block, then it implements `AutoCloseable`; if a type is found in an `implements` clause, then it's changed to an interface; etc.

I haven't made the changes you asked for yet, but I'll make sure to address them before I give you the final version.

theron-wang committed:
add type parameter support

theron-wang committed:
add handling for conditionals + operator in type guessing

theron-wang committed:
some fixes + best effort output

theron-wang committed:
add deleted files

theron-wang committed:
make it build

theron-wang committed:
fix some bugs

theron-wang commented:
Approximately 60 cases pass now. Some of these cases fail because they have slightly different definitions, though.

For example, take `UnsolvedStaticSimpleField`:

Expected:
```java
package org.sampling;

public class Baz {

    public static OrgSamplingBazMyFieldSyntheticType myField;
}
```

Generated:
```java
package org.sampling;
public class Baz {
    public static org.sampling.OrgSamplingBazMyFieldSyntheticType myField;
}
```

The only difference is the FQN, which I don't see any harm in keeping in the output. Do you think we should just change the test cases to reflect this or should we trim the FQN? Trimming the FQN would require a change in the way unsolved symbols are currently outputted, though, since we'd have to pass information from the unsolved type into each unsolved member, in addition to what we currently do: pass information from the unsolved members into the unsolved type (if you look at the `toString` methods this will make more sense).

kelloggm commented:
> The only difference is the FQN, which I don't see any harm in keeping in the output. Do you think we should just change the test cases to reflect this or should we trim the FQN?

Generally, I think output with the FQN is better. So, in this case, I'd argue that the new output is actually an improvement on the old output, and you should definitely change the test case.

Going the opposite direction (FQN -> no FQN) would be a regression, though.

kelloggm commented on src/main/java/org/checkerframework/specimin/JavaParserUtil.java:
```diff
@@ -373,4 +360,101 @@ static String removeMethodReturnTypeAndAnnotationsImpl(String declAsString) {
   }
+
+  /**
+   * Checks to see if an expression (usually a field or method) is static.
```
> usually a field or method

Since these are expressions, this must be a _reference_ to a field or method, rather than a field or method definition.

kelloggm commented on src/main/java/org/checkerframework/specimin/JavaParserUtil.java:
```diff
@@ -373,4 +360,101 @@ static String removeMethodReturnTypeAndAnnotationsImpl(String declAsString) {
+    Expression scope = null;
+
+    if (expr.isNameExpr()) {
+      nameOfScope = nameOfExpr = expr.asNameExpr().getNameAsString();
```
This double-assignment pattern is a bit unintuitive to read and is usually considered bad style in Java. It's more verbose to use three lines for this, but I'd prefer it.

kelloggm commented on src/main/java/org/checkerframework/specimin/JavaParserUtil.java:
```diff
@@ -373,4 +360,101 @@ static String removeMethodReturnTypeAndAnnotationsImpl(String declAsString) {
+        return true;
+      }
+
+      if (!importDecl.isStatic() && importDecl.getNameAsString().endsWith("." + nameOfScope)) {
```
This condition is exactly identical to the condition above, but with `!` before the call to `isStatic()`. That's unexpected, given the method name, and seems suspicious to me. Certainly it justifies an explanatory comment, if it's actually reasonable to do.

theron-wang replied:
There's a slight difference in the two conditions (`nameOfScope` vs `nameOfExpr` at the end of each line), but I did add some comments to make this a bit more clear.

kelloggm commented on src/main/java/org/checkerframework/specimin/JavaParserUtil.java:
```diff
@@ -373,4 +360,101 @@ static String removeMethodReturnTypeAndAnnotationsImpl(String declAsString) {
+   * Checks to see if an expression (usually a field or method) is static.
+   *
+   * @param expr The expression
+   * @return Whether it is static or not
```
Is this method really algorithmic, in the sense that it always gets the right answer? It seems to me that it is based on heuristics, and if that's a fair description it should be in the Javadoc.

kelloggm commented on src/main/java/org/checkerframework/specimin/Slicer.java:
```diff
@@ -0,0 +1,212 @@
+  private final UnsolvedSymbolGenerator unsolvedSymbolGenerator;
+  private final TypeRuleDependencyMap typeRuleDependencyMap;
+
+  public Slicer(
```
This constructor needs Javadoc explaining its arguments and, more importantly, where they come from.

kelloggm commented on src/main/java/org/checkerframework/specimin/Slicer.java:
```diff
@@ -0,0 +1,212 @@
+    this.unsolvedSymbolGenerator = unsolvedSymbolGenerator;
+  }
+
+  public void slice() throws IOException {
```
Is this method _always_ supposed to be called on a member of this class (I suspect so). If that's the case, make the contract simpler: this method can be private, and the constructor can call it.

Alternatively, make the constructor private and provide a static method that both constructs the object and runs the slice() method, and then returns the result to the caller.

kelloggm commented on src/main/java/org/checkerframework/specimin/Slicer.java:
```diff
@@ -0,0 +1,212 @@
+import org.checkerframework.specimin.unsolved.UnsolvedSymbolAlternates;
+import org.checkerframework.specimin.unsolved.UnsolvedSymbolGenerator;
+
+public class Slicer {
```
I would expect a slicer to take a program and a dependency map as input and output the slice. It's not clear that this class actually has that contract, though: the constructor has some other arguments, and the slice() method is void. How does a user of this class get the output? The Javadoc should explain how the class is intended to be used.

kelloggm commented on src/main/java/org/checkerframework/specimin/Slicer.java:
```diff
@@ -0,0 +1,212 @@
+    }
+
+    if (!generatedSymbolSlice.isEmpty()) {
+      // Step 2: Handle the post-processing worklist
```
What is the post-processing worklist? It's not mentioned anywhere; maybe `slice()` needs Javadoc explaining how it differs from the algorithm in the paper?

kelloggm commented on src/main/java/org/checkerframework/specimin/Slicer.java:
```diff
@@ -0,0 +1,212 @@
+    // Step 4: remove all nodes except the ones marked for preservation OR package/import
+    // declarations
+    for (CompilationUnit cu : toSlice.values()) {
+      sliceCompilationUnit(cu);
```
Confusingly-named method, given what the comment above says that it does.

kelloggm commented on src/main/java/org/checkerframework/specimin/Slicer.java:
```diff
@@ -0,0 +1,212 @@
+
+  public void addToWorklist(Node node) {
+    worklist.add(node);
+  }
```
Why do these methods (this one and `getGeneratedSymbolSlice()`) need to be public? They seem like class internals.

kelloggm commented on src/main/java/org/checkerframework/specimin/Slicer.java:
```diff
@@ -0,0 +1,212 @@
+   *
+   * @param c The list of nodes.
+   */
+  private void addToBeginningOfWorklist(List<Node> c) {
```
Rather than this method, we probably should use a different data structure - maybe a FILO queue? 

In general, the need to write a method like this one is almost always an indication ("code smell") that you should do something differently at the architecture or design level - in this case, it indicates that the data structure we chose doesn't quite match the needs of the algorithm.

theron-wang replied:
Yeah, this didn't make me too happy when I wrote it. I changed it around so this method isn't necessary: the target member information goes FILO (worklist does removeLast instead of removing from the beginning), but the type rule dependency map adds to the end, so that part would be FIFO. The reasoning behind this is because certain statements, like A a = new A(), need to be handled before a.method() (so type A is generated). The old method was to circumvent the lack of an addAll to the beginning of the Deque, but doing removeLast resolves the need for such a method. Is this still a problem or do you think it's fine?

kelloggm replied:
> Is this still a problem or do you think it's fine?

That sounds fine. I'm not too concerned about efficiency here, just making sure the code is readable and we don't need to reinvent anything. I'll double-check it when I re-review this code.

kelloggm commented on src/main/java/org/checkerframework/specimin/Slicer.java:
```diff
@@ -0,0 +1,212 @@
+   *
+   * @param node The node to slice
+   */
+  private void sliceNode(Node node) {
```
I'd prefer that we reserve the term "slice" for the overall process, and use a different term for the internal step of removing nodes that this method implements.

kelloggm reviewed (changes requested):
I haven't reviewed the classes in the `unsolved` package yet; I'll try to get to them later.

theron-wang committed:
more bug fixes + test case updates

theron-wang committed:
make requested changes + add docs

theron-wang commented on src/main/java/org/checkerframework/specimin/SpeciminRunner.java:
```diff
@@ -202,217 +216,61 @@ private static void performMinimizationImpl(
-      String primaryTypeQualifiedName = "";
-      if (compilationUnit.getPrimaryType().isPresent()) {
-        // the get() is safe because primary type here is definitely not a local declaration,
-        // which does not have a fully-qualified name.
```
I don't think this handling is removed, at all; it's less explicitly handled now, of course, but I still think the functionality is there. First, the addition a few lines below (`fqnToCompilationUnits.put(declaredClassQualifiedName, compilationUnit);`) makes sure that both non-primary and primary classes point to the same compilation unit. In the type rule dependency map, when a reference to a type is found, it uses the `fqnToCompilationUnits` map to find the right CU and also adds the non-primary class to the slice. The newer code doesn't handle these cases separately.

kelloggm replied:
Okay, I understand. This is a better design, and removes a complication from the algorithm. Nice job!

theron-wang commented on src/main/java/org/checkerframework/specimin/TargetMemberFinderVisitor.java:
```diff
@@ -1,65 +1,43 @@
    * spaces remove for ease of comparison.
    */
-  private final Set<String> targetMethodNames;
+  private final Set<String> targetMethods;
```
Since `SpeciminStateVisitor` no longer exists, and `targetMethods` is no longer unavailable as a field name, I just thought it sounded better to rename it to `targetMethods`. There is no difference between the old field and the current field. I did change it back to `targetMethodNames`, though.

theron-wang committed:
handle array types

theron-wang commented:
@kelloggm How much do you think we need to stick to the old naming conventions with generated field/method types? I'm encountering a few failing tests where the refactor generates shorter names than the old version. For example, `UnsolvedMethodWithArrayParameter` fails because the return type for `UnsolvedType.processArray(intArray);` is differently named. The old has it named `OrgTestingUnsolvedTypeProcessArrayReturnType`, while the new has it named `ProcessArrayReturnType`, both in the same package. If it's not too big of a deal, I won't fix this, but if you think it's important to stick to the original, then I can, too. It definitely is possible to replicate the original behavior in this instance, but I'm not sure if it's worth the effort.

kelloggm commented:
> How much do you think we need to stick to the old naming conventions with generated field/method types?

In general, it's not a strict contract: these generated method/field names are arbitrary. As long as we're confident that the name is unambiguous, it's fine.

> The old has it named OrgTestingUnsolvedTypeProcessArrayReturnType, while the new has it named ProcessArrayReturnType, both in the same package

In this case, I think there is something to worry about: what if there are two methods called `processArray` with different argument lists in the same class, both of which are used by the target? E.g., suppose that we have:
```
processArray(int[] foo) { ...}
```
and
```
processArray(int[] foo, boolean bar) { ... }
```
with calls to each. I don't think the previous code would get this correct, either, but since we're thinking about this we should try our best to make sure that the name will always be unique.

theron-wang committed:
some more bug fixes

kelloggm reviewed (commented):
Partial review of `unsolved`, mostly focused on `FullyQualifiedNameGenerator`. I'll keep going in a bit, need to turn my attention to a different task now.

theron-wang committed:
add ambiguity case for unsolved method calls in solvable method/constructor calls

theron-wang commented on src/main/java/org/checkerframework/specimin/unsolved/FullyQualifiedNameGenerator.java:
```diff
@@ -0,0 +1,984 @@
+   * When evaluating an expression, there is only one possible type. However, the location of an
+   * expression could vary, depending on the parent classes/interfaces of the class which holds the
+   * expression. When determining the location of a {@link FieldAccessExpr}, {@link MethodCallExpr},
+   * or {@link NameExpr}, use this method instead of {@link #getFQNsForExpressionType} because it
```
No, because `#getFQNsForExpressionType` can still be used with those types. This returns the location of the expression, while `#getFQNsForExpressionType` returns its type. If you called the two functions on a MethodCallExpr, `foo()`, for example, the two methods would both be valid but would return different results. `#getFQNsForExpressionType` would return the return type (maybe `FooReturnType`) but `#getFQNsForExpressionLocation` would return the FQNs of potential locations (i.e., static import, ancestor class/interface).

kelloggm replied:
> When determining the location of a {@link FieldAccessExpr}, {@link MethodCallExpr}, or {@link NameExpr}, use this method instead of {@link #getFQNsForExpressionType}

That makes this statement confusing, because that statement implies that substituting the two methods is a sensible thing to do for other kinds of expressions. We should rephrase this to make it clear that these methods return different things (i.e., what you said in your last comment).

theron-wang commented on src/main/java/org/checkerframework/specimin/unsolved/FullyQualifiedNameGenerator.java:
```diff
@@ -0,0 +1,984 @@
+            for (ReferenceType type : unionType.getElements()) {
+              try {
+                // If a type in the union type is resolvable, the location of the expression will
+                // be in a built-in Java superclass. In this case, return an empty map.
```
If a union type is `UnsolvedException | NullPointerException`, then any method called on the NameExpr representing an exception of this type will be in `Exception` or `Throwable`. This was my original thought process and why I left the comment. 

I do realize that in this case, there may be some ambiguity: if it were `UnsolvedException | MyException`, and `MyException` is resolvable, and `MyException` extends some unsolvable class, then it could be in this unsolvable parent class. I'll leave a TODO to address that later.

kelloggm replied:
Please also add the reasoning in your first paragraph to the comment in the code.

theron-wang commented on src/main/java/org/checkerframework/specimin/unsolved/FullyQualifiedNameGenerator.java:
```diff
@@ -0,0 +1,984 @@
+    // If an annotation is @Foo.Bar, we need to find the import with org.example.Foo, not
+    // org.example.Foo.Bar.
+    // Hence, look for the import with the "earliest" scope (with @Foo.Bar, this would be Foo).
+    String getImportedName = anno.getNameAsString();
```
I don't think so, I believe this is the minimum. The method above deals with `ClassOrInterfaceType` which returns another `ClassOrInterfaceType` with the scope, while this method deals with `Name` as the scope. The annotation parent handling is also a little different so that can't be moved into the same method either.

theron-wang committed:
make some changes

kelloggm reviewed (commented):
More comments; I'm now most of the way through `unsolved`, although I didn't look at all of `UnsolvedSymbolGenerator` closely yet.

theron-wang commented on src/main/java/org/checkerframework/specimin/unsolved/UnsolvedClassOrInterfaceAlternates.java:
```diff
@@ -0,0 +1,272 @@
+    UnsolvedClassOrInterfaceAlternates generated;
+
+    if (generatedSymbolsMap.containsKey(packageName)) {
+      return (UnsolvedClassOrInterfaceAlternates) generatedSymbolsMap.get(packageName);
```
> When I read this class' constructor, I (incorrectly, it seems) assumed that the keys to this map were FQNs.

You're right about this assumption.

I think the variable `packageName` is misleading here. Perhaps `qualifier` would be a better name; for inputs like `org.example.MyClass.MyOtherClass`, `org.example.MyClass` would be in `packageName`, but it's clearly not a package name. I also wrote this method before we clarified that we're assuming class paths represent classes, not packages, so the previous implementation treated `org.example.MyClass.MyOtherClass`, `org.example.MyClass`, and `org.example` all as classes. ll update that.

theron-wang commented on src/main/java/org/checkerframework/specimin/unsolved/UnsolvedMethod.java:
```diff
@@ -0,0 +1,217 @@
+    // type of arguments, because the compiler cannot tell them apart. The compiler does not
+    // consider return type when differentiating methods, so you cannot declare two methods with the
+    // same signature even if they have a different return type."
+    return other.name.equals(this.name) && other.parameterList.equals(parameterList);
```
I don't think this is too big of a deal, since most places where equality is necessary deal with `UnsolvedMethodAlternates` instead of `UnsolvedMethod`. The only case, as of right now, where `UnsolvedMethod` equality is used is in the `UnsolvedSymbolEnumerator`. Do you think its acceptable to just leave a note in the Javadoc to use caution with this method for this case?

theron-wang committed:
make some more changes

theron-wang committed:
make comment changes from yesterday's review

theron-wang committed:
make some fixes

theron-wang committed:
make fixes

theron-wang committed:
bug fixes + short review changes

theron-wang committed:
add support for type arguments

theron-wang committed:
must implement methods and other fixes

kelloggm commented on src/main/java/org/checkerframework/specimin/Slicer.java:
```diff
@@ -182,7 +182,9 @@ private void handleElement(Node node) {
         // if there are unresolvable argument types
-        resolved = asResolvable.resolve();
+        if (!(node instanceof FieldDeclaration)) {
+          resolved = asResolvable.resolve();
```
Why is this call to `resolve` skipped if `node` is a `FieldDeclaration`? Adding an explanation to the comment would be helpful.

kelloggm commented on src/main/java/org/checkerframework/specimin/StandardTypeRuleDependencyMap.java:
```diff
@@ -314,4 +339,149 @@ public List<Node> getRelevantElements(Object resolved) {
+        }
+
+        for (ResolvedMethodDeclaration resolvedMethodDecl : methods) {
+          // Don't preserve equals(Object) and hashCode()
```
Why are these two methods, specifically, excluded? I would expect that any justification for excluding `equals` and `hashCode` would also apply to other object methods, like `clone`, no? If these two methods are special for some reason, this comment should explain why.

kelloggm replied:
> equals and hashCode are in the Set interface

But they are already defined on the object because they are defined in `java.lang.Object`, so being in an interface is immaterial. I think this exception should apply to all of the methods in `java.lang.Object`, not just to equals and hashCode. It's okay to just add a TODO though - I don't think this is very important.

kelloggm commented on src/main/java/org/checkerframework/specimin/StandardTypeRuleDependencyMap.java:
```diff
@@ -314,4 +339,149 @@ public List<Node> getRelevantElements(Object resolved) {
+          }
+
+          for (MethodDeclaration methodDecl : originalTypeDecl.findAll(MethodDeclaration.class)) {
+            if (areAstAndResolvedMethodLikelyEqual(resolvedMethodDecl, methodDecl)) {
```
Why can't we use FQNs for the parameters in `areAstAndResolvedMethodLikelyEqual`? That would make this correct beyond a doubt.

kelloggm commented on src/test/resources/constraintType/expected/com/example/Simple.java:
```diff
@@ -1,5 +1,6 @@
 package com.example;
 
+import java.util.Set;
```
`Set` isn't used in this example. While there's not necessarily _harm_ in unnecessary imports, I'd prefer that Specimin not generate them.

Removing unnecessary imports is something that IDEs and formatters already do well, though. Maybe the simplest solution is to run an automated formatter (e.g., [google-java-format](https://github.com/google/google-java-format)) over Specimin's output automatically?

kelloggm reviewed (commented):
This review is only on the changes in commit c369d72

kelloggm reviewed (commented):
Review of commit 6e928e0

kelloggm commented on src/test/resources/methodreturnfullyqualifiedgeneric/expected/com/bar/Bar.java:
```diff
@@ -2,7 +2,7 @@
 public class Bar<T> {
 
-    public static com.bar.Bar<com.example.InOtherPackage2> getOtherPackage2() {
+    public static com.bar.Bar<com.foo.InOtherPackage2> getOtherPackage2() {
```
https://github.com/njit-jerse/specimin/issues/422

kelloggm commented on src/test/resources/unsolvedmethodtypecorrect/expected/com/example/Simple.java:
```diff
@@ -1,15 +1,13 @@
-
-    void baz() {
-        throw new java.lang.Error();
+        final LocalVariables locals = f.getLocals();
```
This change doesn't look correct to me. The original input file is:
```
package com.example;

import org.example.LocalVariables;
import org.example.Foo;

class Simple {
    // Target method.
    void bar() {
        Foo f = new Foo();
        f.getLocals().set(0);
        baz();
    }

    void baz() {
        // To trigger JavaTypeCorrect.
        Foo f = new Foo();
        final LocalVariables locals = f.getLocals();
    }
}
```

Putting the contents of the `baz()` method into the `bar()` method definitely seems incorrect.

kelloggm replied:
> I modified the test case (the input as well) to move the line into bar() so that we'd still be able to test for the same thing, which is why the output here looks the way it does.

Thanks, I missed that you'd changed both.

kelloggm reviewed (commented):
Review of changes in commit 81648d5

theron-wang commented:
@kelloggm If you're not reviewing already, you can hold off on doing so for the next day or so; I'm going to try and get rid of all the miscellaneous bugs before I tackle some of the larger problems. I expect that once I'm done with these bug fixes, non-UnsolvedSymbolGenerator files will stay relatively unchanged, so you shouldn't have to review another time.

theron-wang committed:
pr review changes

theron-wang committed:
fix some bugs

theron-wang commented:
@kelloggm Ready for review, except for UnsolvedSymbolGenerator. I bumped up the gradle version to try to get it to run faster, but to no avail (30s to 60s still for each change).

kelloggm commented on src/main/java/org/checkerframework/specimin/StandardTypeRuleDependencyMap.java:
```diff
@@ -133,6 +144,54 @@ public List<Node> getRelevantElements(Node node) {
+          }
+        } catch (UnsolvedSymbolException ex) {
+          // No preservation of super() call necessary because the synthetic type will
+          // have a default constructor.
```
(we should just add this test case, I think)

theron-wang replied:
The case that you pointed out does indeed work, but you did help me realize another problem:

```java
class Foo extends Bar {
    public Foo() {
       super("any argument", "some other argument");
    }

    public static Bar bar() {
        Foo foo = new Foo();
        return new Bar("a consistent argument");
    }
}
```

This slightly modified version (where `bar()` is the only target method) doesn't preserve the super(...) call and that super constructor is not preserved. Maybe we should make it so if the parent is unresolvable, then always preserve these super calls? Or should we always explicitly generate a default constructor?

I think my original code was under the assumption that the parent class would not have an object creation expression in the target method, so the generated parent would have no constructor (and thus have the default constructor).

kelloggm replied:
> I think my original code was under the assumption that the parent class would not have an object creation expression in the target method, so the generated parent would have no constructor (and thus have the default constructor).

I don't think we can make this assumption, as your test case shows (nice adaptation, by the way!).

> Maybe we should make it so if the parent is unresolvable, then always preserve these super calls?

I prefer this approach. It may result in us over-preserving constructors, but I'd rather do that than generate default constructors that might not actually be present in the original version of the code. At least we know that these super constructors must exist, because we have legitimate calls to them in code from the user.

kelloggm commented on src/main/java/org/checkerframework/specimin/unsolved/FullyQualifiedNameGenerator.java:
```diff
@@ -376,21 +378,61 @@ else if (expr.isCastExpr()) {
+        FullyQualifiedNameSet rightType =
+            getFQNsForExpressionTypeImpl(binary.getRight(), canRecurse);
+
+        // Remaining operators only work with primitive/String types
```
What about boxed types like `java.lang.Integer`? Those are also possible here, although maybe you can just treat them as their primitive equivalents.

kelloggm commented on src/main/java/org/checkerframework/specimin/unsolved/FullyQualifiedNameGenerator.java:
```diff
@@ -376,21 +378,61 @@ else if (expr.isCastExpr()) {
+        }
+
+        // TODO: alternates here? cases here are all primitives (except for +, which could be
+        // String)
```
It's easy to cover all of the alternates here (we know the list of primitives), so we should just do so to avoid a headache later.

kelloggm replied:
> I can do this after I address the other failing cases, though.

It's not a priority, I agree. Leaving the TODO here is fine: we can come back to it or not later.

kelloggm commented on src/main/java/org/checkerframework/specimin/unsolved/UnsolvedFieldAlternates.java:
```diff
@@ -50,12 +87,40 @@ public static UnsolvedFieldAlternates create(
+  public void updateFieldTypesAndMustPreserveNodes(
+      Map<MemberType, CallableDeclaration<?>> typesToPreserveNodes) {
+    // Update in-place; intersection = removing all elements in the original set
+    // that isn't found in the updated set
```
Can we have one implementation of this intersection logic that's shared by the different `XAlternates` classes?

kelloggm commented:
> @kelloggm Ready for review, except for UnsolvedSymbolGenerator. I also bumped up the gradle version to try to get it to run faster, but to no avail (30s to 60s still for each change).

I've reviewed everything except `UnsolvedSymbolGenerator.java`.

theron-wang committed:
fix anonymous classes, override in synthetic types, pr changes, lambda crashes in progress

theron-wang committed:
stop crashing for cases with solvable lambda constraint types

theron-wang commented on src/main/java/org/checkerframework/specimin/JavaParserUtil.java:
```diff
@@ -766,14 +862,51 @@ public static List<MethodDeclaration> tryResolveMethodCallWithUnresolvableArgume
 
+  /**
+   * Given an enum constant declaration, returns all possible constructors that match the arity and
+   * known types of the arguments.
```
I didn't change it to null, but I did clarify with this statement in the Javadoc: This returns an empty list if no matching constructors were found or if the declaring type could not be solved.

Is that alright? The `tryResolve` methods are different in purpose than the other `try` methods so I think it shouldn't be a big deal for consistency if it returns null or not (it would make it a bit more annoying for the method's user if it were null instead of empty list).

kelloggm commented on src/main/java/org/checkerframework/specimin/StandardTypeRuleDependencyMap.java:
```diff
@@ -609,7 +632,37 @@ private void getAllMustImplementMethodsImpl(
+        // then the composed map should be E --> String.
+        List<Pair<ResolvedTypeParameterDeclaration, ResolvedType>> typeParametersMap;
+
+        // Now, substitute previous type parameters through the current map
```
Do we have a test case that requires this behavior? If not, we should write one to double check that this is correct. (I recall that this wasn't in the last commit, so if adding it caused a test to go from failing to passing, we're already ok. If not, though, you should write a corresponding test.)

kelloggm replied:
Nevermind, I found MustImplementMethodsComplexTest

kelloggm commented on src/main/java/org/checkerframework/specimin/unsolved/UnsolvedSymbolGenerator.java:
```diff
@@ -988,6 +999,121 @@ private void handleConstructorCall(
+        switch (specifier) {
+          case PUBLIC -> "public";
+          case PROTECTED -> "protected";
+          case PRIVATE -> "private";
```
The `PRIVATE` case isn't possible, right, since this method must be an override? I would throw here rather than proceeding if we do hit this case, because I think it's an error condition.

kelloggm reviewed (commented):
Review of 7a6fc1c

kelloggm commented on src/main/java/org/checkerframework/specimin/JavaParserUtil.java:
```diff
@@ -64,6 +76,35 @@ private JavaParserUtil() {
+  /**
+   * Set this TypeSolver instance to be used by the JavaParserFacade, so we don't have to find it
+   * through reflection in JavaParserSymbolSolver. This field is initialized once in SpeciminRunner.
+   * Note that by the time you evaluate the value of this field, it should already be non-null.
```
The [MonotonicNonNull annotation](https://checkerframework.org/api/org/checkerframework/checker/nullness/qual/MonotonicNonNull.html) exactly expresses this contract.

kelloggm replied:
(you'll also want an `@EnsuresNonNull` annotation on `setTypeSolver`, I think)

kelloggm commented on src/main/java/org/checkerframework/specimin/JavaParserUtil.java:
```diff
@@ -1365,10 +1406,269 @@ public static String getInitializerRHS(String variableType) {
+
+      // Don't use a map here: in case a ResolvedType is the same for different parameters, we could
+      // have different pieces of information (i.e., two parameters are both resolved type T, but
+      // the args
```
nit: awkward formatting

kelloggm commented on src/main/java/org/checkerframework/specimin/JavaParserUtil.java:
```diff
@@ -1365,10 +1406,269 @@ public static String getInitializerRHS(String variableType) {
+   * @param expression The expression to find the declaration of (method, field)
+   * @return The resolved object, if it exists; null otherwise
+   */
+  public static @Nullable Object tryFindCorrespondingDeclarationForConstraintQualifiedExpression(
```
I didn't read this that closely and will need to review this again when I have a bit more mental energy available. I don't see anything obviously wrong for now, though.

kelloggm reviewed (commented):
Quick review of commit 02a8153

theron-wang committed:
small fixes and add additional test cases based on #423

theron-wang committed:
super type relationships + bug fixes

theron-wang commented:
@kelloggm Ready for review, including UnsolvedSymbolGenerator

theron-wang committed:
allow inheritance for bounded wildcards and other type arguments + javadoc comments

theron-wang commented:
I know you're away at the moment, but whenever you get a chance again, you can just review whatever I have at the moment.

theron-wang committed:
fix a few bugs

kelloggm reviewed (commented):
I wrote these review comments on Friday and forgot to post them. I'll try to review the current status of the PR now as best I can, time permitting.

kelloggm reviewed (commented):
Review of commit ece78bb

kelloggm reviewed (commented):
Review of changes in commit 07bc8ea

theron-wang commented on src/main/java/org/checkerframework/specimin/unsolved/UnsolvedSymbolGenerator.java:
```diff
@@ -2310,6 +2179,236 @@ else if (node instanceof MethodCallExpr methodCall) {
+   * <ul>
+   *   <li>The method declaration is unsolved and has a synthetic definition for it
+   *   <li>The unsolved declaring type of this method has a solvable child class with the same
+   *       method signature
```
Never mind, I think this test case matches the question I'm asking here: http://github.com/njit-jerse/specimin/tree/main/src/test/resources/voidreturndouble

Overrides are not preserved, but the return types are preserved; void is used instead of a synthetic return type in the test case linked above.

theron-wang committed:
address pr comments and start adding method reference support

kelloggm reviewed (commented):
review of changes in commit ebd54b3

theron-wang committed:
method references

theron-wang committed:
bug fixes + javadoc

theron-wang committed:
fix must implement methods

theron-wang committed:
must implement methods should generate in unsolved class + private outer classes should make a method parameter be java.lang.Object

theron-wang committed:
Merge branch 'main' of https://github.com/njit-jerse/specimin into refactor

theron-wang committed:
include methods that are implemented but referenced by an abstract super call, account for abstract overrides, and fix anonymous class handling

theron-wang committed:
handle cases where the erased type is solvable but the type args are not

theron-wang committed:
make return type of unsolved method a type arg if it must extend a final class

theron-wang committed:
Merge branch 'main' of https://github.com/njit-jerse/specimin into refactor

theron-wang committed:
handle comparison expressions better and fix a bug

theron-wang committed:
handle cases where an unsolvable expression in a lambda expression method call argument prevents resolution of an otherwise-solvable method

theron-wang committed:
remove logs that i missed

theron-wang committed:
fix cases like MismatchTypeParameterHardCaseTest

theron-wang committed:
fix crashes when handling type parameter uses in anonymous classes

theron-wang committed:
handle crashes relating to the resolution of expressions that have an unbounded wildcard as their type

theron-wang committed:
fix mismatched type parameter names in overridden method generations

theron-wang committed:
change functional interface parameter types in unsolved method generations to use method-scoped type variables instead of wildcards

theron-wang committed:
fix anonymous class method resolution and parent method return type/parameter types

theron-wang committed:
fix cases where methods with the same erasure require type variables + upgrade gradle

theron-wang committed:
upgrade all dependencies to latest versions and apply intellij refactor suggestions

theron-wang committed:
fix MethodAmbiguityException bug introduced in JavaParser 3.28.2

theron-wang commented:
@coderabbitai configuration

theron-wang committed:
temporarily change coderabbit config to see why it's not reviewing

theron-wang commented:
@coderabbitai full review

theron-wang committed:
temporarily block most paths from review

theron-wang commented:
@coderabbitai full review

theron-wang committed:
maybe now?

theron-wang commented:
@coderabbitai full review

theron-wang committed:
add all paths

theron-wang committed:
flip flags back on

theron-wang commented:
@coderabbitai full review

theron-wang committed:
revert all coderabbit changes and fix uncompilable test case

theron-wang committed:
make it so `./gradlew build` succeeds

theron-wang committed:
try one more time for coderabbit

theron-wang committed:
coderabbit does not work: revert

theron-wang commented:
@kelloggm Ready for re-review, with some notes:
* Most file changes (200+ in test files) since your last review are import changes from upgrading JUnit from 4 to 6
* CF-3032 no longer crashes, but it takes much longer (>100s) compared to other evaluation cases
* CF-3025 is now preserved
* MethodRefUndefTest (the one you added sometime this fall/winter) doesn't pass and is ignored for now
* For CI results, I canceled the last two workflows so you can just look at the third-to-last run (the one that passed) for the most recent logs
* All alternates are not generated (only best effort right now)
* I tried running CodeRabbit earlier, but I think there are too many files in this PR so I am not going to run it (525 files vs limit of 150)

kelloggm commented on src/main/java/org/checkerframework/specimin/unsolved/UnsolvedMethodAlternates.java:
```diff
@@ -0,0 +1,531 @@
+
+/**
+ * /** Given a name, return type, a set of parameters and a set of potential encapsulating classes,
+ * this class allows for alternates of a same field to be generated in different locations. If a
```
nit: there's a typo here - this should say "alternates of the same **method**", not "field"

kelloggm commented on src/main/java/org/checkerframework/specimin/unsolved/UnsolvedSymbolGenerator.java:
```diff
@@ -0,0 +1,4347 @@
+   *
+   * @param fqnsToCompilationUnits A set of fully-qualified names to compilation units
+   */
+  @SuppressWarnings("method.invocation")
```
nit: When making a warning suppression like this, you should also use the checker key so that it applies only to the checker whose output you intended to suppress. It's also a good idea to write a comment explaining why the suppression is okay - that is, why the warning was a false positive. In this case, since this code was written like a year ago, I think you can just leave a TODO comment.

kelloggm commented on src/main/java/org/checkerframework/specimin/SpeciminRunner.java:
```diff
@@ -245,17 +257,20 @@ public void run() {
    * @param disableRootValidation whether to disable root validation
    * @throws IOException if there is an exception
    */
+  @SuppressWarnings("UnusedVariable") // Remove once ambiguityResolutionPolicy is used
```
Should this be removed now? I think probably yes; otherwise I'm concerned about why we have `AmbiguityResolutionPolicy` in this PR at all.

kelloggm reviewed (commented):
A few low-level comments on files that I happened to look at, but this was by no means a thorough review. I think we should merge this PR based on high-level metrics (like "it preserves everything in specimin-evaluation") rather than based on the low-level details, because I'm confident that this PR _does_ [improve the overall health of the codebase](https://google.github.io/eng-practices/review/reviewer/standard.html) and a low-level review will not be pleasant for anyone: it's too much code to go through, and @theron-wang didn't write most of it recently.

I do think we should do a few small experiments to make sure that this change doesn't cause serious regressions, and I am concerned that the new ambiguity resolution policies don't have dedicated tests. So, I propose that we do the following before we merge:
* write a few dedicated tests for each of the two new ambiguity resolution policies. These can be adaptations of existing tests (with different expected outputs), and need not be comprehensive - the standard I'm looking for here is "if we write code in the future that fundamentally breaks one of the other policies, we'll notice in CI". As far as I can tell, there aren't any new tests that use the new ambiguity resolution policies, but if I've missed them in this huge PR just let me know @theron-wang and point me at them.
* run a few experiments to be sure that this change doesn't make the percentage of arbitrary methods in arbitrary projects that compile after being Speciminified significantly worse. There is already infrastructure in the repo to do this (we had it running as a CI step for a while, but it's a bit fragile/noisy for that use case). In particular, the file [.github/workflows/check_compilable_percentage.yml](https://github.com/njit-jerse/specimin/blob/main/.github/workflows/check_compilable_percentage.yml) has logic to run this kind of experiment on a GitHub runner, and overall it should be a pretty straightforward activity. Pick a few repos (3 smallish ones of your choice is probably sufficient - I suggest picking ones on the order of 10-20k LoC) and run every method in each through Specimin both before and after this change. As long as the % of results that compile is _not worse_ in the condition with this PR, we can merge it. Let me know if you have any questions or concerns about this plan.

theron-wang commented:
@kelloggm I forgot about the two ambiguity resolution policies; that's also why you also don't see any tests relating to it in this PR. Do you want me to add it, or should I just remove it altogether?

kelloggm commented:
> I forgot about the two ambiguity resolution policies; that's also why you also don't see any tests relating to it in this PR

If you haven't actually implemented either of them, it's better to cut the option. I'd like to at least support the "all" ambiguity resolution policy, though - I think it will actually be useful in some cases. But, if it's not actually implemented yet, it's okay if that's in a follow-up PR. If that's what you're going to do, please make an issue so that we don't forget about it (in case you don't get to it this summer).

theron-wang committed:
remove ambiguity resolution policy, address CR comments, add repos to CI_repository_list.csv

theron-wang committed:
change CI repo to one with src/main/java

theron-wang commented:
@kelloggm 

I ran `check_compilable_percentage_CI` on these repos:
- https://github.com/M3CHR0M4NC3R/plume-util-static
- https://github.com/cbeust/jcommander
- https://github.com/jopt-simple/jopt-simple

CI run on current: https://github.com/theron-wang/specimin/actions/runs/28538258310/job/84605098564 (this is on my personal fork; if you can't see this and would like to, let me know)
CI run on refactor: https://github.com/njit-jerse/specimin/actions/runs/28535005216/job/84594018661

| Metric | Current | Refactor |
| --- | --- | --- |
| Full successes | 897 | 869 |
| Failed minimizations | 24 | 39 |
| Failed compilations | 61 | 74 |
| Validate accuracy | 20 | 39 |

Validate accuracy is higher on the refactor? I don't know why

kelloggm commented:
@theron-wang these results look fine. Not great - there are slightly more failures with the refactor than without it - but close enough that I don't think the regressions will be problematic. It would be great to try to fix some of these (or at least check if there is some low-hanging fruit), but that doesn't have to happen as part of this PR.

> Validate accuracy is higher on the refactor? I don't know why

I'm also surprised by that, but it's not important - validate accuracy is a relative measure compared to a baseline, and the baseline in the repo is probably out of date (since we're not using this CI step anymore). You can safely ignore it.

kelloggm reviewed (approved):
Once the "check compilable percentage" CI logic has been removed again I think this is ready to merge. Thanks for your hard work on this big rewrite!

theron-wang committed:
revert CI_repository_list.csv

theron-wang commented:
Sounds good -- I removed the repositories I added in yesterday and turned off the workflow through GitHub (as it was yesterday as well). I have turned on auto-merge so it should merge in a few minutes.
